### PR TITLE
jit: support for clz, ctz and popcnt

### PR DESCRIPTION
**Pull request by mathetake**
- [x] clz
- [x]  ctz
- [x] popcnt

### CODE REVIEW

File: wasm/jit/jit_amd64.go [hunk 1]
```diff
@@ -876,7 +876,8 @@ func (c *amd64Compiler) compileMulForInts(is32Bit bool, mulInstruction obj.As) e
 
 	// Now we have the result in the AX register,
 	// so we record it.
-	c.locationStack.pushValueOnRegister(resultRegister)
+	result := c.locationStack.pushValueOnRegister(resultRegister)
+	result.setRegisterType(generalPurposeRegisterTypeInt)
```
Comment (mathetake) [inline wasm/jit/jit_amd64.go:880]: this was missing from #111 so I fixed it while there

**Comment by mathetake** [comment]
lol CLZ is only failing on macos 😄

**Comment by mathetake** [comment]
Looks like we have to implement something like this only for mac (from https://developer.apple.com/documentation/apple-silicon/addressing-architectural-differences-in-your-macos-code)
```
int cnt(int x){
    int count;
    if(x == 0){
        return (sizeof(x) * CHAR_BIT);
    }
    __asm__ (
        "bsrl %1, %0\n\t" // undefined when operating on 0
        "xorl $0x1f, %1"
        : "=r" (count)
        : "r" (x)
    );
    return count;
}
```

**Comment by mathetake** [comment]
I hope https://github.com/tetratelabs/wazero/pull/113/commits/898a8b639a608363b9f9e90916a99d00b80e1172 fixed for mac.

**Comment by mathetake** [comment]
now only failing for the case where the target value is zero

**Comment by mathetake** [comment]
ok realized that now CLZ succeeded on mac but the only failing case is for CTZ

**Comment by mathetake** [comment]
ready for review!

File: wasm/jit/jit_amd64.go [hunk 2]
```diff
@@ -916,6 +918,224 @@ func (c *amd64Compiler) compileMulForFloats(instruction obj.As) error {
...
+		cmpZero.To.Offset = 0
+		c.addInstruction(cmpZero)
+
+		jmpIfNonZero := c.newProg()
+		jmpIfNonZero.As = x86.AJNE
+		jmpIfNonZero.To.Type = obj.TYPE_BRANCH
+		c.addInstruction(jmpIfNonZero)
+
+		// If the value is zero, we just push the const value.
+		ifZeroConst := c.newProg()
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64.go:1067]: can this be extracted to a function to reuse? eyes are squinting, but looks the same as above zero case.

File: wasm/jit/jit_amd64.go [hunk 3]
```diff
@@ -916,6 +918,224 @@ func (c *amd64Compiler) compileMulForFloats(instruction obj.As) error {
...
+	countBits.From.Type = obj.TYPE_REG
+	countBits.From.Reg = target.register
+	countBits.To.Type = obj.TYPE_REG
+	countBits.To.Reg = target.register
+	if o.Type == wazeroir.UnsignedInt32 {
+		countBits.As = x86.APOPCNTL
+	} else {
+		countBits.As = x86.APOPCNTQ
+	}
+	c.addInstruction(countBits)
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64.go:1131]: I bet you are happy macOS didn't mess with you again!

File: wasm/jit/jit_amd64_test.go [hunk 4]
```diff
@@ -2636,6 +2639,315 @@ func TestAmd64Compiler_compileMul(t *testing.T) {
 	})
 }
 
+func TestAmd64Compiler_compilClz(t *testing.T) {
+	t.Run("32bit", func(t *testing.T) {
+		for _, tc := range []struct{ in, exp uint32 }{
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64_test.go:2644]: I  know you are tired of typing, but maybe dodge "exp" as later my eyes will cross when we have exponents in some instructions..

```suggestion
		for _, tc := range []struct{ input, expected uint32 }{
```

Comment (codefromthecrypt) [reply wasm/jit/jit_amd64_test.go:2644]: or even expectedBitCount as I kindof looked at one of the later functions sideways a second forgetting I am seeing hex on one side and bit count on the other

**Review by codefromthecrypt — APPROVED** [review]
Good detective work. No massive comment this time as I can't find anything that isn't explained well enough!

File: wasm/jit/jit_amd64_test.go (hunk 4 above)
Comment (codefromthecrypt) [reply wasm/jit/jit_amd64_test.go:2644]: I guess it changes what we are expecting, so maybe expectedUnsetBitCount vs expectedSetBitCount could be the more pedantically ideal name :D

Comment (mathetake) [reply wasm/jit/jit_amd64_test.go]: https://github.com/tetratelabs/wazero/pull/113/commits/396cd5a5d2313a0f09f1e7c5dea0eba8635cf562 done!

File: wasm/jit/jit_amd64.go (hunk 2 above)
Comment (mathetake) [reply wasm/jit/jit_amd64.go]: https://github.com/tetratelabs/wazero/pull/113/commits/9c906ab4e5290f407e5a5dc6d55291f3463f73d6 there you go!

**Comment by mathetake** [comment]
Thanks for the look!

**Pull request merged**